src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs

```diff
@@ -394,13 +400,72 @@ GenericParameterValue genericParam
 					} else {
 						requiredMemberTypes = GetDynamicallyAccessedMemberTypesFromBindingFlagsForMembers (bindingFlags);
 					}
+					if (!MemberTypesAreUnsupported (memberTypes)) {
+						DynamicallyAccessedMemberTypes requiredMemberTypesMaskFromMemberTypes;
+						requiredMemberTypesMaskFromMemberTypes = GetDynamicallyAccessedMemberTypesFromMemberTypesForMembers (memberTypes);
+						requiredMemberTypes &= requiredMemberTypesMaskFromMemberTypes;
+					}
 
 					var targetValue = _annotations.GetMethodThisParameterValue (calledMethod, requiredMemberTypes);
 
 					// Go over all types we've seen
 					foreach (var value in instanceValue.AsEnumerable ()) {
-						// Mark based on bitfield requirements
-						_requireDynamicallyAccessedMembersAction.Invoke (value, targetValue);
+						if (value is SystemTypeValue systemTypeValue) {
+							foreach (var stringParam in argumentValues[0].AsEnumerable ()) {
+								if (stringParam is KnownStringValue stringValue && !BindingFlagsAreUnsupported (bindingFlags) && !MemberTypesAreUnsupported (memberTypes)) {
+									// determine if we've got a prefix (for example, abc* searches for anything starting with abc)
+									var isPrefix = stringValue.Contents.EndsWith("*");
+									if (isPrefix)
+									{
+										// Mark based on bitfield requirements
+										_requireDynamicallyAccessedMembersAction.Invoke (value, targetValue);
+										continue;
+									}
+									var name = stringValue.Contents;
+
+									// we may not need to search for constructors depending on the value of 'name', since they're always named '.ctor' (II.10.5.1)
+									DynamicallyAccessedMemberTypes requiredMemberTypes2 = requiredMemberTypes;
+									if (name != ".ctor") {
+										requiredMemberTypes2 &= ~(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors);
+									}
+
+									AddReturnValue (MultiValueLattice.Top); // Initialize return value (so that it's not autofilled if there are no matching members)
+
+									// search for all the things we want by name
+									if ((requiredMemberTypes2 & (DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)) != 0) {
```

Comment:
Why not something like `if (name == ".ctor" && ...)` and remove the check above about constructors - would probably simplify this a little bit.